### PR TITLE
Add a fedmsg consumer that handles libraries.io messages

### PR DESCRIPTION
**Pull request by jeremycline**
This adds a fedmsg consumer that handles messages from libraries.io by
way of the sse2fedmsg bridge. This is not complete yet, but it mostly works and
I wanted some initial feedback to see what people thing. Things left to do are:

- Tests

- A way to handle when libraries.io announces a version with a prefix (like v) and Anitya doesn't. 

- What to do when anitya finds a different version than libraries.io

- What do do when anitya doesn't find any version at all

This also sets up the Vagrant environment to use this.

### CODE REVIEW

File: anitya/lib/librariesio_consumer.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        name = librariesio_msg['name']
+        platform = librariesio_msg['platform']
+        try:
+            ecosystem = list(filter(
+                lambda eco: eco.librariesio_name == platform, ECOSYSTEM_PLUGINS.get_plugins()))[0]
+        except IndexError:
+            # TODO probably better to raise an exception here
+            return
+
+        session = Session()
```
Comment (pypingou) [inline anitya/lib/librariesio_consumer.py]: Should it be closed somewhere?

Comment (jeremycline) [inline anitya/lib/librariesio_consumer.py]: Based on my (not yet complete) understanding of SQLAlchemy, the whole point of using its ``scoped_session`` is to create a session registry where each thread gets its own session so you don't have to pass session objects around. The session "created" here is the same session "created" [here](https://github.com/release-monitoring/anitya/pull/433/files#diff-5379d281c0b6a0060c136e2c384cd573R133). When ``session.remove()`` is called the session is removed from the registry and its resources released. Next time ``Session()`` is called a new session is placed in the thread-local registry.

Not really on-topic, but we're not using a lot of nice SQLAlchemy features. I think in the near future we need to alter all the database APIs to use custom query classes (where appropriate) in combination with the [query_property](http://docs.sqlalchemy.org/en/latest/orm/contextual.html#sqlalchemy.orm.scoping.scoped_session.query_property) rather than all those static methods on the models that accept sessions. The backend should create a scoped session rather than relying on the Flask application do to so (otherwise we should be using flask-sqlalchemy since we're tying the backend to the frontend anyway).

File: anitya/lib/librariesio_consumer.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            Project: the Anitya project that maps to the librariesio project;
+                this may be a new project.
+            None: If there is no ecosystem that maps to the librariesio platform
+        """
+        librariesio_msg = message['body']['msg']['data']
+        name = librariesio_msg['name']
+        platform = librariesio_msg['platform']
+        try:
+            ecosystem = list(filter(
+                lambda eco: eco.librariesio_name == platform, ECOSYSTEM_PLUGINS.get_plugins()))[0]
```
Comment (ncoghlan) [inline anitya/lib/librariesio_consumer.py]: This would likely be clearer as a search loop (which would also save creating the list every time):

    for ecosystem in ECOSYSTEM_PLUGINS.get_plugins():
        if ecosystem.librariesio_name == platform:
            break
    else:
        # Log a debug message about an unknown ecosystem
        # (that will be pretty common - libraries.io currently tracks more ecosystems than Anitya does)
        return

File: anitya/lib/librariesio_consumer.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        # TODO handle things like 'v' prefix. Should be a general version normalizer
+        # used everywhere.
+        # TODO handle when we find a different version than librariesio. Create the project anyway?
+        equivalent_versions = (librariesio_version == anitya_version)
+        if not equivalent_versions:
+            _log.info('libraries.io found %r had a latest version of %s, but Anitya found %s',
+                      project, librariesio_version, anitya_version)
+        return equivalent_versions
+
+    def _fedmsg_to_project(self, message):
```
Comment (ncoghlan) [inline anitya/lib/librariesio_consumer.py]: I'd suggest calling this `_fedmsg_to_candidate_project_update`, as it does more than just work out which project the message relates to:

- for an unknown project, it creates a candidate project instance
- for an existing project, it speculatively updates the latest version info
- in both cases, it *avoids* committing the change, until after the legitimacy of the info can be checked

File: anitya/lib/librariesio_consumer.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        This serves two purposes. Firstly, we connect to libraries.io over HTTP
+        so we don't have any authentication. Secondly, we might map the
+        libraries.io project to the wrong Anitya project, so this is a good
+        opportunity to catch and log those problems.
+        """
+        librariesio_version = message['body']['msg']['data']['version']
+        try:
+            anitya_version = get_version(project)
+        except exceptions.AnityaPluginException as e:
+            # TODO what to do here? Bail? Trust libraries.io?
```
Comment (ncoghlan) [inline anitya/lib/librariesio_consumer.py]: Bailing out here with a logged error is correct, as the query may fail because the candidate project was complete nonsense, and instead injected via an MitM attack against the unauthenticated feed from libraries.io.

File: anitya/lib/ecosystems/crates.py [hunk 5]
```diff
@@ -28,3 +28,4 @@ class CratesEcosystem(BaseEcosystem):
 
     name = 'crates.io'
     default_backend = 'crates.io'
+    librariesio_name = 'Cargo'
```
Comment (ncoghlan) [inline anitya/lib/ecosystems/crates.py]: Having to make all the ecosystem plugins libraries.io aware irks me a bit, but I guess the alternative is making the libraries.io handling code aware of all the ecosystems, and that runs somewhat counter to the design of the plugin model.

However, normalising with `str.lower()` would be sufficient to map all of the libraries.io names except `cargo`, and for that we could potentially change the ecosystem name on the Anitya plugin side.

Alternatively (and to better support deciding we want to change canonical ecosystem names in the future), this could be handled by way of a general purpose "aliases" field on the plugins, rather than a libraries.io specific field.

Comment (jeremycline) [inline anitya/lib/ecosystems/crates.py]: Yeah I'm not pleased with this either, it was just the first thing that popped into my head. Having an "aliases" field definitely sounds better to me.

File: anitya/lib/librariesio_consumer.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        else:
+            _log.debug('Dropped librariesio update to %s for %s (%s) since it is on the '
+                       'unsupported %s platform', version, name, homepage, platform)
+            return
+
+        session = Session()
+        project = Project.by_name_and_ecosystem(session, name, ecosystem.name)
+        if project is None:
+            # We've found a new project!
+            project = Project(
```
Comment (jeremycline) [inline anitya/lib/librariesio_consumer.py]: A problem with this approach that I've just found is that we're not going to be emitting the correct fedmsgs since those are sent all over the place via ``anitya.log``.

I could re-work this PR to use that approach, but I'm inclined to open a new PR that adds [ORM event listeners](http://docs.sqlalchemy.org/en/latest/orm/events.html) that publish the messages instead.

**Comment by jeremycline** [comment]
Okay, I think this is ready.

**Comment by ralphbean** [comment]
https://github.com/fedora-infra/sse2fedmsg

^^ cool.

File: anitya/tests/test_config.py [hunk 7]
```diff
@@ -163,6 +165,7 @@ def test_full_config_file(self, mock_exists, mock_log):
             'SOCIAL_AUTH_LOGIN_URL': '/login/',
             'SOCIAL_AUTH_LOGIN_REDIRECT_URL': '/',
             'SOCIAL_AUTH_LOGIN_ERROR_URL': '/login-error/',
+            'CREATE_LIBRARIESIO_PROJECTS': True,
```
Comment (ncoghlan) [inline anitya/tests/test_config.py]: Should this be `'CREATE_LIBRARIESIO_PROJECTS': create_librariesio_projects` to make sure the two test settings always stay in sync?

File: anitya/tests/test_librariesio_consumer.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+                    },
+                    'event': 'event',
+                    'id': None
+                }
+            }
+        }
+        self.mock_hub = mock.Mock(config={'environment': 'dev', 'topic_prefix': 'anitya'})
+
+    def test_dev_environment_topics(self):
+        """Assert when the environment is set to prod, only the prod topic is used."""
```
Comment (ncoghlan) [inline anitya/tests/test_librariesio_consumer.py]: This docstring is incorrect for the dev test case.

File: anitya/tests/test_config.py (hunk 7 above)
Comment (jeremycline) [reply anitya/tests/test_config.py]: I'm not entirely sure what you mean here, could you rephrase this or show me what you mean? Thanks!

Comment (ncoghlan) [reply anitya/tests/test_config.py]: Further up in this file, you have the line `create_librariesio_projects = True`. Presumably that global and this setting are supposed to be kept consistent, so the code in my comment was meant to be a literal replacement for the current dict entry.

Alternatively, if the global isn't actually used anywhere (I didn't check), it could be deleted entirely.

Comment (jeremycline) [reply anitya/tests/test_config.py]: Ahhh I see. I think the diff is not very clear. That's inside a string used as the test configuration file's content.

Comment (ncoghlan) [reply anitya/tests/test_config.py]: Ah, you're right - I only looked at the diff, not the full original file. In that case, cool, it makes sense as is :)
